Monitor the queue and acquire logs

Jullianos opened the pull request.

vitalied commented on Gemfile:
```diff
@@ -49,6 +49,10 @@ gem 'sentry-ruby'
 gem 'sentry-sidekiq'
 
+gem 'faraday'
+gem 'faraday_middleware'
```
@Jullianos @AndriiMysko `faraday_middleware` is deprecated.
Why you are not using `rest-client` instead of `faraday`?

Jullianos replied:
I've decided to go with `faraday_middleware` because that was the gem used in travis-api from where I took the RemoteLog code.

vitalied replied:
@Jullianos you can't decide by yourself. Every library used should be approved by our side: See 5.4 in PRD.
Gem `faraday_middleware` is deprecated. We are not going to use unsupported code in new applications. Pls use `rest-client`

Jullianos replied:
I've refactored the code and removed faraday_middleware and faraday as dependencies.

vitalied reviewed (changes requested):
@Jullianos @AndriiMysko use `rubocop -A`

vitalied commented on app/services/process_logs_service.rb:
```diff
@@ -16,22 +16,159 @@ def call
       Rails.logger.error(e.message)
     end
   end
+class ProcessLogsService < BaseLogsService
+  def initialize(log_ids)
+    super()
+
+    @log_ids = log_ids
+  end
+
+  def call
+    Rails.logger.info("Received log_ids: #{@log_ids.inspect}")
+
+    begin
+      Travis::Lock.exclusive('process_logs', lock_options) do
+        update_log_entries
+      end
+    rescue => e
+      Rails.logger.error(e.message)
+    end
+  end
```
@Jullianos @AndriiMysko what this code is doing here?

Jullianos replied:
Based on the requirements once the process job gets picked up from the queue it should fetch the logs by the provided log_ids, update their status to :started and create new ScanTrackerEntry rows, and finally it should write the log content into a local directory/file from where it will be picked up and censored.

vitalied replied:
@Jullianos pls review the whole file. You have duplicated code after rebase here.

Jullianos replied:
Yes, stupid mistake, thanks for noticing.

Jullianos commented on lib/travis/remote_log.rb:
```diff
@@ -0,0 +1,350 @@
+      @remote ||= Travis::RemoteLog::Remote.new(platform: platform)
+    end
+
+    class Client
```
One of the task requirements was that in case the log is archived in the S3 bucket (the content is not present in the database) we should make a specific call using the remote log from travis-api to the S3 bucket and retrieve the archived log content.

vitalied replied:
@Jullianos @AndriiMysko @speedywizard if you already using S3 client lib to upload files directly to AWS S3, why you are not using the same S3 client lib to get the files?

Jullianos replied:
In regards to this I've already cleared the remote_log.rb file from all the unnecessary code with my latest commit.

Jullianos commented on Gemfile:
```diff
@@ -49,6 +49,10 @@ gem 'sentry-ruby'
 
+gem 'faraday'
+gem 'faraday_middleware'
+gem 'virtus'
```
Again I've decided to go with `virtus` because this was the gem of choice for the travis-api RemoteLog class.

vitalied replied:
@Jullianos you can't decide by yourself. Every library used should be approved by our side: See 5.4 in PRD.
`virtus` was discontinued in 2019. We can't use it in new projects.
You should use `dry-types`, `dry-struct` and `dry-schema` instead.

Jullianos replied:
I've refactored the code and removed virtus as a dependency.

Jullianos commented:
> `A`

@AndriiMysko you might have to help me here as I had troubles configuring rubocop locally today. I will try to set it up tomorrow again, but just in case.

AndriiMysko commented:
@vitalied comments and rubocop issues were fixed

vitalied commented on lib/travis/remote_log.rb:
```diff
@@ -0,0 +1,232 @@
+    end
+
+    class Remote
+      extend Forwardable
+
+      def_delegators :archive_client, :fetch_archived_url, :fetch_archived_log_content
```
@Jullianos @AndriiMysko why this is needed?

AndriiMysko commented on app/services/process_logs_service.rb:
```diff
@@ -20,18 +20,73 @@ def call
-    return unless logs.exists?
-
-    log_ids = logs.pluck(:id)
+    logs_query = Log.where(id: @log_ids, scan_status: :queued)
+    log_ids = logs_query.pluck(:id)
+    return if log_ids.empty?
```
@vitalied Previous code failed the tests. `exists?` return statement always returned even if there were results

Jullianos commented:
With my latest commit I've removed all the redundant code from remote_log and replaced fog with aws-sdk-s3. But for it to work as expected it requires that at least one of these is configured properly:
- Environment variables: ENV['AWS_ACCESS_KEY_ID'] and ENV['AWS_SECRET_ACCESS_KEY']
- Programmatically: Aws.config[:credentials]
- The shared credentials ini file at ~/.aws/credentials
In my opinion the environment variables is the safest approach but I'm willing to make modifications if needed.

vitalied commented:
> With my latest commit I've removed all the redundant code from remote_log and replaced fog with aws-sdk-s3. But for it to work as expected it requires that at least one of these is configured properly:
> 
> * Environment variables: ENV['AWS_ACCESS_KEY_ID'] and ENV['AWS_SECRET_ACCESS_KEY']
> * Programmatically: Aws.config[:credentials]
> * The shared credentials ini file at ~/.aws/credentials
>   In my opinion the environment variables is the safest approach but I'm willing to make modifications if needed.

@Jullianos yea, env vars is the best option.
Pls do this in an intializer

vitalied reviewed (changes requested):
@Jullianos specs are needed
